1-Draft/RFCXXXX-Trap-Handle-Non-Terminating-Errors.md

```diff
@@ -0,0 +1,126 @@
+---
+RFC: RFC0XXX
+Author: Tyler Leonhardt
+Status: Draft
+Area: Traps
+Comments Due: 
+---
+
+# Trap should handle non-terminating errors
+
+This RFC proposes that we change the behavior of the Trap statement to also trap non-terminating exceptions.
+
+To do this,
+we will also add a property on ErrorRecord to determine if it was a terminating or non-terminating exception.
+
+With this change, the following code:
+
+```pwsh
+trap {
+    "hello"
+    continue
+}
+Get-Process IDontExist
+```
+
+will output the same as the following code that works today:
+
+```pwsh
+trap {
+    "hello"
+    continue
+}
+Get-Process IDontExist -ErrorAction Stop
+```
+
+If you wanted to handle both terminating and non-terminating errors in your
+`trap`
+you can use the
+`ErrorType`
+property:
+
+```pwsh
+trap {
+    if($_.ErrorType -eq "Terminating") {
+        "oh no!"
+    } else {
+        "this is ok"
+        continue
+    }
+}
+```
+
+## Motivation
+
+The Trap statement has fallen out of favor in recent years to it's error handling cousin -
+the try/catch block.
+
+This gives us an opportunity to revamp the Trap statement so that it can serve a purpose that the try/catch could not serve.
+
+The purpose would be to offer a
+*single solution*
+to handle both terminating and non-terminating issues.
+
+## Specification
+
+`ErrorRecord`
+will get an additional property called
+`ErrorType`
+this will come from an
+`enum`
+called
+`ErrorRecordType`
+which has 2 values:
+
+* `Terminating`
+* `NonTerminating`
+
+Traps will then also be passed non-terminating errors to be handled. 
+
+## Alternate Proposals and Considerations
+
+### Introduce a `trapall`
```

Comment:
My personal preference is `trapall`, since:
- It doesn't break the existing behaviour of `trap`
- It isn't a syntactic break; `trapall` will still parse and execute, but will just throw when the statement invoking `trapall` is hit in older PowerShell verisons. Compare this to `trap -NonTerminating` which will not parse in older PowerShell versions. This means you can wrap the former in a version condition and it will still work cross-version.

We might even able to carefully define `trapall` so that `ICustomAstVisitor` isn't broken, by reusing the `TrapStatementAst` type with a new parameter in the constructor detailing whether it's a `trap` or a `trapall` statement.